Peg clock value to whisper timestamp

cammellos opened the pull request:
This commit pegs the clock value to max +120 seconds from the whisper
timestamp.
In this way we avoid having a client increase arbitrarily the chat clock value.

Validating that the clock value is not too much in the past would be useful, but currently that would break `datasync` (depending on the range), as `datasync` re-transmit the original message in a new whisper message, so it's a case where the difference between the two is ok, but it also means that we are exposed to malicious insertion of messages in the past (there's no way to know if a message has been genuinely re-transmitted by datasync or the user maliciously inserted a message in the history).

One  way to mitigate this is to have a UX solution, whereby those messages are for example marked and displayed with a warning.

cammellos committed:
Peg clock value to whisper timestamp

This commit pegs the clock value to maximum + 120 seconds from the whisper
timestamp.
In this way the we avoid the scenario where a client makes the timestamp
increase arbitrarely.

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (20)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 98af6bbe | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1804/1/) | 2020-01-17 14:12:59 | ~48 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200117-141211-98af6b.zip) |
| :heavy_check_mark: | 98af6bbe | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1804/1/) | 2020-01-17 14:17:38 | ~5 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200117-141209-98af6b.zip) |
| :heavy_check_mark: | 98af6bbe | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1804/1/) | 2020-01-17 14:20:20 | ~8 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200117-141207-98af6b.aar) |
| | | | | | | |
| :heavy_check_mark: | 0a40764c | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1804/2/) | 2020-01-20 07:26:47 | ~56 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200120-072550-0a4076.zip) |
| :heavy_check_mark: | 0a40764c | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1804/2/) | 2020-01-20 07:30:59 | ~5 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200120-072549-0a4076.zip) |
| :heavy_check_mark: | 0a40764c | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1804/2/) | 2020-01-20 07:32:11 | ~6 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200120-072548-0a4076.aar) |
| | | | | | | |
| :x: | 72970196 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1804/3/) | 2020-01-20 13:26:10 | ~20 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1804/3/consoleText) |
| :heavy_check_mark: | 72970196 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1804/3/) | 2020-01-20 13:26:50 | ~57 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200120-132552-729701.zip) |
| :heavy_check_mark: | 72970196 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1804/3/) | 2020-01-20 13:28:53 | ~3 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200120-132551-729701.zip) |
| | | | | | | |
| :heavy_check_mark: | 23d805dc | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1804/4/) | 2020-01-20 13:28:17 | ~44 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200120-132732-23d805.zip) |
| :heavy_check_mark: | 23d805dc | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1804/4/) | 2020-01-20 13:31:53 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200120-132855-23d805.zip) |
| | | | | | | |
| :heavy_check_mark: | 99dd4b5a | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1804/5/) | 2020-01-20 13:30:44 | ~40 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200120-133003-99dd4b.zip) |
| :heavy_check_mark: | 99dd4b5a | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1804/5/) | 2020-01-20 13:34:16 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200120-133155-99dd4b.zip) |
| :heavy_check_mark: | 99dd4b5a | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1804/4/) | 2020-01-20 13:35:36 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200120-132959-99dd4b.aar) |
| | | | | | | |
| :heavy_check_mark: | f51bb2b2 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1804/6/) | 2020-01-20 13:33:37 | ~40 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200120-133257-f51bb2.zip) |
| :x: | f51bb2b2 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1804/5/) | 2020-01-20 13:35:52 | ~13 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1804/5/consoleText) |
| :heavy_check_mark: | f51bb2b2 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1804/6/) | 2020-01-20 13:36:56 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200120-133419-f51bb2.zip) |
| | | | | | | |
| :heavy_check_mark: | af6aef50 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1804/7/) | 2020-01-20 13:37:57 | ~38 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200120-133718-af6aef.zip) |
| :heavy_check_mark: | af6aef50 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1804/7/) | 2020-01-20 13:39:40 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200120-133717-af6aef.zip) |
| :heavy_check_mark: | af6aef50 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1804/6/) | 2020-01-20 13:42:31 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200120-133715-af6aef.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 3845ef1f | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1804/8/) | 2020-01-20 14:07:41 | ~46 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200120-140655-3845ef.zip) |
| :heavy_check_mark: | 3845ef1f | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1804/8/) | 2020-01-20 14:09:24 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200120-140653-3845ef.zip) |
| :heavy_check_mark: | 3845ef1f | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1804/7/) | 2020-01-20 14:14:52 | ~8 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200120-140651-3845ef.aar) |
| | | | | | | |
| :heavy_check_mark: | c859908a | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-1804/9/) | 2020-01-20 16:28:49 | ~38 sec | `linux` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-desktop-200120-162811-c85990.zip) |
| :heavy_check_mark: | c859908a | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-1804/9/) | 2020-01-20 16:30:40 | ~2 min | `ios` | [:package:`zip`](https://status-go.ams3.digitaloceanspaces.com/status-go-ios-200120-162808-c85990.zip) |
| :heavy_check_mark: | c859908a | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-1804/8/) | 2020-01-20 16:33:38 | ~5 min | `android` | [:package:`aar`](https://status-go.ams3.digitaloceanspaces.com/status-go-android-200120-162807-c85990.aar) |

adambabik commented on protocol/chat.go:
```diff
@@ -207,39 +207,39 @@ func oneToOneChatID(publicKey *ecdsa.PublicKey) string {
 }
 
-func OneToOneFromPublicKey(pk *ecdsa.PublicKey) *Chat {
+func OneToOneFromPublicKey(pk *ecdsa.PublicKey, timesource clockValueTimesource) *Chat {
```
This is a public method so for the convenience `clockValueTimesource` should be also public.

adambabik commented on protocol/message_validator.go:
```diff
@@ -2,17 +2,52 @@ package protocol
 	"strings"
 
 	"github.com/status-im/status-go/protocol/protobuf"
+	"github.com/status-im/status-go/protocol/v1"
```
We are in `github.com/status-im/status-go/protocol`. This is not needed.

cammellos replied:
How do I import `MembershipUpdateMessage` otherwise? I tried removing it, but without it wont' compile, is there another way? (in `messenger.go` we import I  believe)

adambabik replied:
Ah sorry. It is needed. I though I deleted this comment...

adambabik commented on protocol/message_validator.go:
```diff
@@ -2,17 +2,52 @@ package protocol
 
+	c := new(big.Int).SetUint64(clock)
+	w := new(big.Int).SetUint64(whisperTimestamp)
+	result := new(big.Int).Sub(c, w)
```
If we only subtract lower from higher, we should never overflow and we wouldn't need to use `big.Int`, right?

adambabik commented on protocol/message_validator.go:
```diff
@@ -2,17 +2,52 @@ package protocol
+	for _, e := range message.Events {
+		// We only compare in one direction for membership update as they are relayed
+		// without the original whisper timestamp
+		if e.ClockValue > timeNowMs && e.ClockValue-timeNowMs > maxWhisperDriftMs {
```
```suggestion
		if err := validateClockValue(e.ClockValue, timeNowMs); err != nil {
```

adambabik commented on protocol/messenger.go:
```diff
@@ -1307,7 +1304,7 @@ func (m *Messenger) SendChatMessage(ctx context.Context, message *Message) (*Mes
 	}
 
-	err := extendMessageFromChat(message, chat, &m.identity.PublicKey)
+	err := extendMessageFromChat(message, chat, &m.identity.PublicKey, m.transport)
```
I would recommend to have a getter on `Messenger`:
```suggestion
	err := extendMessageFromChat(message, chat, &m.identity.PublicKey, m.getTimeSource())
```
so that when the implementation changes, you will only need to change it in one place, `getTimeSource` method. It looks odd that we pass `transport` to `extendMessageFromChat` which is a helper.

cammellos commented:
@adambabik addressed feedback, 

I actually had to change validating that the clock value is not too much in the past as that would break datasync, an explanation is above.

Thanks!

pedropombeiro commented on protocol/message_validator_test.go:
```diff
@@ -72,13 +87,15 @@ func (s *MessageValidatorSuite) TestValidateRequestAddressForTransaction() {
-		Message protobuf.ChatMessage
+		Name             string
+		WhisperTimestamp uint64
+		Valid            bool
```
I know this is not directly related to this PR (other than the fact that this PR added more error types), but instead of true/false, should we compare with the actual expected error message?

cammellos replied:
I don't mind either approaches, basically `false` here is to say: `This is not to be validated anymore`, while `Error` means that it can be retried. Probably not in this PR, but you think matching the errors explicitly is a better way? If so I can push the change in a separate PR.
thanks!

adambabik commented on protocol/transport/whisper/whisper_service.go:
```diff
@@ -360,6 +361,10 @@ func (a *WhisperServiceTransport) Track(identifiers [][]byte, hash []byte, newMe
 	}
 }
 
+func (a *WhisperServiceTransport) GetCurrentTime() uint64 {
```
Let's add a comment or a unit test that makes sure it returns unix timestamp in milliseconds.
